diff --git a/swift/utils/build-script-impl b/swift/utils/build-script-impl
index ff451cfb43f..bda4485c744 100755
--- a/swift/utils/build-script-impl
+++ b/swift/utils/build-script-impl
@@ -1371,6 +1372,8 @@ function swift_c_flags() {
         echo -n " -fno-stack-protector"
     fi
 
+    #echo -n " --Wl,-z,max-page-size=16384"
+
     local host=$1
 
     case $host in
@@ -2331,7 +2331,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     continue
                 fi
 
-                if [[ "${BUILD_SWIFT_TOOLS}" == "0" ]]; then
+                if [[ "${CROSS_COMPILE_BUILD_SWIFT_TOOLS}" == "0" ]]; then
                     echo "Skipping building Foundation Macros for ${host}, because the host tools are not being built"
                     continue
                 fi
@@ -2920,7 +2920,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     continue
                 fi
 
-                if [[ "${BUILD_SWIFT_TOOLS}" == "0" && "${product}" == "foundation_macros" ]]; then
+                if [[ "${CROSS_COMPILE_BUILD_SWIFT_TOOLS}" == "0" && "${product}" == "foundation_macros" ]]; then
                     echo "Skipping installing Foundation Macros for ${host}, because the host tools are not being built"
                     continue
                 fi
diff --git a/swift/utils/swift_build_support/swift_build_support/targets.py b/swift/utils/swift_build_support/swift_build_support/targets.py
index fba09416ddb..9eab6cf1c78 100644
--- a/swift/utils/swift_build_support/swift_build_support/targets.py
+++ b/swift/utils/swift_build_support/swift_build_support/targets.py
@@ -165,6 +165,7 @@ class AndroidPlatform(Platform):
 
         flags += '-sdk %s/sysroot ' % (android_toolchain_path)
         flags += '-tools-directory %s/bin' % (android_toolchain_path)
+        #flags += '-Xlinker --Wl,-z,max-page-size=16384'
         return flags
 
     def cmake_options(self, args):
